### ISSUE TITLE
Adding a "PointSource" function

### ISSUE THREAD

**Issue by cschreib** [perwin/imfit#4]
Hello Peter. I was wondering what is currently the best way to fit a point source to an image. Would it make sense to add a "PointSource" function where only the position and total flux could be varied? This seems like a basic component that is, as far as I know, currently missing.

**Comment by perwin** [comment]
The way I and several other people have done it is to use a circular Gaussian with very small, fixed size. E.g.,

    FUNCTION Gaussian
    PA      0   fixed
    ell      0   fixed
    I_0     100   0.1,1e5
    sigma   0.1   fixed

(where you would adjust the initial value and possible limits on I_0 as seems sensible for the particular source.)

**Comment by cschreib** [comment]
That is what I suspected :) The problem, I guess, it that the meaning of I_0 is not immediately clear to me in this situation. I suppose it can always be inferred from the total flux using the formal integral of a 2D Gaussian, but this seems rather counter-intuitive; I would much rather work directly with the total flux, which is easier to guess.

I've also had trouble getting the fit to converge when using this solution, but I have still to investigate this further; it may just be a problem of the data.

**Comment by perwin** [comment]
Hmm.... I_0 is the peak flux of the Gaussian. To my mind, it's easier to guess the I_0 value by starting with peak observed flux in the central pixel of a putative point source.

But... that's why interactions like this are interesting and useful: they expose blind spots in what I think are "obvious" or "intuitive". If you think it would be useful to have a "point-source" function where the only parameter is the total flux, then I can look into making something like that (basically, a circular Gaussian with small, fixed sigma).

**Comment by cschreib** [comment]
Unless I misunderstood what you suggested, I don't think that taking I_0 as the peak observed flux works. To begin with, the observed image is convolved with the PSF. If the source is truly point-like, then its peak flux on the image will be the total flux of the source multiplied by the peak value of the PSF. So one would first need to divide the observed peak flux by the PSF's peak value. Then, this gives the total flux, and not I_0. At fixed total flux, I_0 depends on the chosen value of sigma (and ell, but let's assume that any sane person would set this to zero). The total flux of the 2D Gaussian is ```I_tot = I_0*2*PI*sigma^2```. This means the right guess for I_0 is ```peak_flux/(2*PI*sigma^2)/PSF_peak```. Not the most convenient in my opinion :) In contrast, one can measure the total flux rather easily by summing pixels within a large enough aperture (in DS9 for example).

Other points to consider:
 - Galfit has it. People migrating from Galfit may be confused if a dedicated point source model does not exist.
 - A point source is simpler to generate than a 2D Gaussian, and this can lead to performance improvements. It should only amount to something like:
```c++
double PointSource::GetValue( double x, double y )
{
  double oversampling = 1.0;
  long x_diff = std::abs(long(x*oversampling) - long(x0*oversampling));
  long y_diff = std::abs(long(y*oversampling) - long(y0*oversampling));
  return x_diff == 0 && y_diff == 0 ? I_tot : 0.0;
  // If the user provided an oversampled PSF then you would need
  // to change the value of "oversampling" above to reflect this.
}

double PointSource::TotalFlux( )
{
  return I_tot;
}
```

Given this, I do think that adding the point source would be a good thing.

**Comment by perwin** [comment]
Re: estimating I_0 from the peak pixel flux: what I was trying to do was describe how I'd generate an initial guess for the fitting process -- an order-of-magnitude estimate for a value to put in the config file for imfit. (You're quite right that this would be inaccurate; I was thinking about how to quickly answer the question "Should I start with I_0 = 100? 1000? 1e6?".)

**Comment by perwin** [comment]
OK, I think the main problem with the PointSource implementation as you've presented it is that it produces point sources with discretized coordinates: they are always located at the center of a pixel, so the chi^2 landscape will be step-shaped in the X0 and Y0 directions. (There's also the lesser problem that "oversampling" is basically an extra parameter of the function, and would have to be correctly specified as a (fixed) parameter in the configuration file.)

Some tests indicate that the step-shaped nature of the PointSource chi^2 landscape completely confuses the Levenberg-Marquardt solver, so that it always settles on the initial X0,Y0 guess as the "best-fit" result. I think this is because if the L-M solver starts off with steps in X0 and Y0 that are < 1 pixel in size, the resulting chi^2 is *exactly the same*, and so it effectively decides it's already at the minimum in the X0-Y0 subspace of the chi^2 landscape.[*] The Nelder-Mead solver doesn't do much better: it doesn't decide that the initial X0,Y0 values are perfect, but it doesn't really do any better at finding the correct X0,Y0 values.

On the other hand, using a narrow, fixed-width Gaussian as the fitting function gives consistent and not terribly inaccurate results, both with L-M and with N-M.

(For the tests, I started with a large circular Moffat image, then downsampled it 5x5, added a small amount of Poisson noise, then fit the resulting "data" image using either the fixed-narrow-Gaussian model or your PointSource model, convolved with the 5x5-downsampled equivalent of the original Moffat function. I repeated this with two more images where the original Moffat was offset by 0.5 to 1.5 pixel amounts in the orignal, full-size image.)

[*] And in fact the default initial step size inside the L-M algorithm is about 1.0e-8 times the initial parameter value.

**Comment by cschreib** [comment]
I see. Doing the mathematically correct thing might involve changing the architecture of the classes a bit, so that function objects can have a say in how their model gets convolved by the PSF. Regular function objects would use the current Convolver, while the PointSource function object would use a different approach where the PSF image is painted directly (since FFT would be overkill in this situation) but interpolated at the current sub-pixel coordinates of the source (e.g., using bicubic interpolation).

This may turn out to be not as efficient after all, but it would at least give the most accurate answer (which may be important when de-blending two close point sources, for example). As a user I would be concerned about "not terribly inaccurate results" :) How inaccurate did it get with the Gaussian?

**Comment by perwin** [comment]
I'd have to rework the main image generation, in the sense of adding an extra stage after the FFT convolution where the point sources were added. The image function for a PSF of this type would have to be rather different, with the ability to be initialized with a PSF image.

In the limited tests I did, the three X0,Y0 values were:
true X0,Y0 -- best-fit X0,Y0
    20.4,20.4 -- 20.472,20.471
    20.5,20.5 -- 20.501,20.501
    20.75,20.75 -- 20.559,20.558

I'll admit I hadn't checked to see how accurate the total flux of the resulting "point source" was when I made that statement above. Looks like the calculated Gaussian flux is almost twice the correct value, which is not very good.

**Comment by perwin** [comment]
OK, I have a preliminary version of a PointSource function which applies bicubic spline interpolation to the PSF image, and the first tests are promising...

**Comment by cschreib** [comment]
Great! Looking forward to try it.

**Comment by perwin** [comment]
Can I assume that you would *not* want to use one of your "do not normalize" interfermetric PSFs for something like this? (Otherwise I'm not sure how to specify the "total flux" parameter...)

**Comment by cschreib** [comment]
Actually I would. For a point source, in this situation the total flux is equal to the peak flux, so I would want a PSF normalized to peak flux of one.

**Comment by cschreib** [comment]
Let me rephrase this more clearly. I would expect that internally the program would simply rescale the PSF (whatever it is) by a factor F. If the PSF is normalized to unit sum (the default), then F is the total flux and is equal to the sum of all the pixels on the source on the image. If I disable the PSF renormalization, then F is an arbitrary scaling factor, and it's up to me to choose the PSF's normalization so that this value makes sense. For example, in case of interferometric images F would still be the total flux if the PSF is normalized to a peak value of one (note that, in this case, the total flux is no longer the sum of the pixel values, but that's how these images work).

**Comment by perwin** [comment]
Would it be possible for me to examine a sample PSF of this type? I still have no idea what one looks like, and it might make it easier for me to understand some of the issue.

**Comment by perwin** [comment]
P.S. I have some vague plans to make an alternate version of the PointSource function that uses Lanczsos3 (and maybe also Lanczos2) interpolation instead of bicubic.

**Comment by cschreib** [comment]
Here is zipped FITS file of a PSF I've used recently: [alma-dirty-beam.fits.zip](https://github.com/perwin/imfit/files/1367079/alma-dirty-beam.fits.zip)
This is the "dirty beam" (i.e., the PSF) of some ALMA observations. You'll notice how the peak value is one, and how the integral of the PSF oscillates above and below zero when going at large radii.

**Comment by perwin** [comment]
Thanks!
(Wow, that's kind of psychedelic...)

**Comment by perwin** [comment]
If you want to give the new PointSource function (using bicubic interpolation) a try, the current version of the code on Github has it. (It's missing some sanity checks, such as making sure you supplied a PSF image.) The current parameters look like this:

    FUNCTION PointSource
    I_tot

The interpolated PSF image is multiplied by the I_tot value, and it should respect the --no-normalize flag.

You do need to have version 2.0 or later of the GNU Scientific Library (I think the current version is 2.3 or 2.4, but some older Linux distributions -- e.g., Ubuntu 14.04 -- only have version 1.6).

**Comment by perwin** [comment]
Hi Corentin,

Just wondering if you'd had a chance to give the new PointSource function a try...

**Comment by cschreib** [comment]
Oops sorry, I didn't see the notification for your reply. I'll test this right away!

**Comment by cschreib** [comment]
It does not work: 
```
../../imfit -c imfit_config.dat  image.fits --psf psf.fits  --save-residual residual.fits --noise noise.fits 
	Image file = image.fits
	configuration file = imfit_config.dat
	noise image = noise.fits
	PSF image = psf.fits
	output residual (data - best-fit model) image = residual.fits
Reading data image ("image.fits") ...
naxis1 [# pixels/row] = 101, naxis2 [# pixels/col] = 101; nPixels_tot = 10201
Reading noise image ("noise.fits") ...
Reading PSF image ("psf.fits") ...
naxis1 [# pixels/row] = 101, naxis2 [# pixels/col] = 101; nPixels_tot = 10201
Function: PointSource
** ERROR: PointSource function specified, but no PSF image was supplied!
Function: PointSource
** ERROR: PointSource function specified, but no PSF image was supplied!
Function: PointSource
** ERROR: PointSource function specified, but no PSF image was supplied!
9 total parameters
Model Object: 10201 data values (pixels)
ModelObject: mask vector applied to weight vector. (10201 valid pixels remain)
Setting up parameter information array ...
9 free parameters (10192 degrees of freedom)
Estimated memory use: 11161433 bytes (10.6 MB)

Performing fit by minimizing chi^2 (data-based errors):
Calling Levenberg-Marquardt solver ...
Segmentation fault (core dumped)
```

Looking at the code that triggers this error, I see that it requires ```psfInterpolator_allocated == true``` before ```AddFunction()``` is called. However, the interpolator is only allocated in ```AddPSFVector()``` if ```pointSourcesPresent == true```, which is set... in ```AddFunction()```. There is a circular logic here.

**Comment by perwin** [comment]
Hah -- well, it *did* work previously, before I cleverly added the pointSourcesPresent check which, as you noted, creates some perfectly circular logic of the Catch-22 kind...

I'll post a comment when I have a fix (I'd like to avoid automatically creating PsfInterpolator objects whenever PSF images are used unless an actual PointSource image-function is being used, but that will require a little thought as to how to handle that.

**Comment by perwin** [comment]
OK, I think the newest commit should be usable -- can you give it a try?

**Comment by cschreib** [comment]
Yes it works now! Thanks. I have run some quality checks, the results of which are quite satisfying.

The tests I made consisted of two point sources of equal flux, separated by some distance on the X axis. I varied this distance on a fine grid, ranging from a fraction of a pixel up to ~five pixels. I did not add any noise since I just wanted to check how accurate the interpolation/fitting could get in an ideal setup.

I performed the test using three different Gaussian PSFs, the width (in pixels) was chosen to match images properties of common instruments: Hubble ACS (0.1" PSF, 0.06"/pixel), Spitzer IRAC (1.7" PSF, 0.6"/pixel), and most ground-based data (0.7" PSF, 0.15"/pixel). This resulted in PSFs of FWHM equal to 1.7, 2.8, and 4.7 pixels, respectively.

I created the input images and PSFs manually, sampling the 2D Gaussian function with a x9 oversampling for each pixel, to obtain quasi-perfect profiles. Then I let imfit try to fit them using the LM algorithm, providing a PSF with the same pixel size as the image (I haven't tried using an oversampled PSF, but I'm not sure: is it actually supported for point sources?).

I did the test twice: once with all parameters free (X0, Y0, and I_tot), and a second time by fixing the positions to their true values and only fitting the flux (I_tot).

The results are shown just below. The first source, always located at the center of the image, is shown on the left, and the second source, which is progressively moved along the X axis, is shown on the right. The bottom panel shows the difference between the true value and the best-fit value, while the top panel shows the absolute value of this difference, in log scale. The curves of different colors correspond to the different PSFs.

**With all parameters free**

![screenshot-area-2018-01-24-115805](https://user-images.githubusercontent.com/2236577/35328821-ed2031ba-00fd-11e8-8929-08a3eefc565c.png)
![screenshot-area-2018-01-24-115814](https://user-images.githubusercontent.com/2236577/35328829-eff7fefe-00fd-11e8-936a-511bee837ca8.png)

**Fitting only the flux, fixed positions to true values**

![screenshot-area-2018-01-24-115823](https://user-images.githubusercontent.com/2236577/35328842-f96feece-00fd-11e8-82f8-f13cd8732fd3.png)

**Results of tests**

- When the sources are separated by > 1 pixel, imfit can find the correct positions with a worst-case precision of 1e-3 pixel, and the flux with a worst-case precision of 0.1% (or 0.01% when the position is fixed). This is pretty good.
- The error is maximal when the source is located exactly between two pixels, and it goes to zero when the source is located exactly at the center of a pixel. The error is also a strong function of the PSF FWHM, with narrower PSFs leading to larger errors. This is the behavior I would expect for interpolation errors.
- When the sources are separated by < 1 pixel and the positions are fixed to their true values, the flux is still recovered with a worst-case precision of 0.1%, which is good.
- When the positions are free to vary and the separation is < 1 pixel, imfit starts to struggle. Typically, one source is given all the flux, and the other one is lost. This happens at separations of 0.8 to 0.6 pixels, depending on the width of the PSF: a wider PSF allows the sources to be better separated.

Nothing I did not expect here, I think it is performing very well. If anything, it shows the good in using images with a pixel size that generously over-samples the PSF, whenever possible.

**Comment by perwin** [comment]
Great! Very nice analysis. (More in a bit; I'm a bit busy with prepping OBs for ESO observations...)

**Comment by perwin** [comment]
To answer your question above ("I haven't tried using an oversampled PSF, but I'm not sure: is it actually supported for point sources?") -- yes, version 1.6 of Imfit does allow oversampled PSF images for the PointSource function (it's required if you specify oversampled regions).

**Comment by perwin** [comment]
I'm going to go ahead and close this issue, but feel free to re-open it if you need to.
